### DISCUSSION TITLE
Multiple-threaded synthesis

### DISCUSSION THREAD

**Discussion by lvtao-sec** [cvc5/cvc5#10490] [Q&A]
Hi,

I'm using cvc5 sygus to synthesize programs.
However, due to the large enumeration space, the speed is too slow.

Are there any options to enable multiple threads to concurrently find the solutions?

I know there is an option `compute-partitions`, but it looks like it doesn't improve the speed.

Thanks in advance.

**Comment by lvtao-sec** [comment]
Hi @ajreynol @mdeters,
I am really sorry to bother you.
Do you know if CVC5 provides options for parallel program synthesis? For example, given one sygus input, multiple threads will enumerate the syntax tree.

I'd appreciate it if you give me some clues. Thanks in advance.

**Comment by ajreynol** [comment]
`compute-partitions` is for parallel SMT solving only.

At the moment there is not a way to leverage parallelism for SyGuS solving. We will keep you posted if this changes.

Btw, there are 2 forms of enumeration in the SyGuS solver (see https://homepage.divms.uiowa.edu/~ajreynol/cav19b.pdf).
These can be enabled by:
`--sygus-enum=smart` and `--sygus-enum=fast`

I would recommend trying both of these, since it varies from application to application which one is better.

Also, you can use `-o sygus` to see the stream of enumerated candidate solutions to further understand how the solver is working.

**Comment by lvtao-sec** [comment]
Thank you, Andrew.
Yes, I'm aware of the smart and fast enumerations and looking forward to the changes.

I plan to modify the cvc5 code to see if I can realize the parallel enumerations. For example, a process/thread takes one syntax tree branch to enumerate at specific points.

Best,
Tao
